### DISCUSSION TITLE
matchPackageTypes

### DISCUSSION THREAD

**Discussion by LucasDemea** [renovatebot/renovate#24888] [Suggest an Idea]
### What would you like Renovate to be able to do?

It would be useful to have an option to filter package types, e.g. composer's `"type": "wordpress-plugin"`.

### If you have any ideas on how this should be implemented, please tell us here.

Not familiar enough with renovate codebase.

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
Can you point to an example project using these types? Typically this would fit into our `depType` field but I see we use that for composer already with `require` and `require-dev`. One possibility is that we support multiple `depTypes`, e.g. `"depTypes": ["require", "wordpress-plugin"]`

**Comment by LucasDemea** [comment]
You can have a look at [roots/bedrock](https://github.com/roots/bedrock).

It's a custom wordpress stack where plugin are managed as composer dependencies having the `wordpress-plugin` type.

Typically, you would install a plugin with e.g. `composer require wpackagist-plugin/akismet` and the [composer/installers](https://github.com/composer/installers) package would handle the installation in the correct wordpress sub-folder (not `vendor`) based on the custom `wordpress-plugin` type.

The composer/installers page shows [a list](https://github.com/composer/installers#current-supported-package-types) of others packages types examples.

**Comment by rarkins** [comment]
How/where are types declared? For example it's not in https://github.com/roots/bedrock/blob/f4a3185251566708a4fa14f71bff879d7831aca1/composer.json#L38

I'm trying to understand how Renovate would learn this type, if it's not from `composer.json`

**Comment by LucasDemea** [comment]
I think it could be confusing to use the same option to specify "depTypes" and "packageTypes", as this is not represented by the same object in `composer.json`. The "packageType" is declared in the dependency's `composer.json` `type` field, while the "depType" is declared in the project main `composer.json`.

e.g.
Main project:
```json
{
    "name": "myproject/myproject",
    "type": "project",
    "require": {
        "some/dependency": "^1.0.0"
    }
}
```

Dependency:
```json
{
    "name": "some/dependency",
    "type": "custom-type"
}
```

**Comment by LucasDemea** [comment]
You can use any type you want. There is no list of predefined types. [composer/installers](https://github.com/composer/installers) makes use of specific types, but you could use any type if you wanted to implement a custom installer, for example.

**Comment by LucasDemea** [comment]
I don't think Renovate could learn them. It would need to be able to use any arbitrary value.

**Comment by rarkins** [comment]
So the type can only be learned after downloading the dependency details from packagist?

**Comment by LucasDemea** [comment]
Idk if it's useful but with `composer show` you can get the type of an installed package (present in lock file).

e.g. `composer show wpackagist-plugin/better-wp-security --format json` :
```json
{
    "name": "wpackagist-plugin/better-wp-security",
    "description": null,
    "keywords": [],
    "type": "wordpress-plugin",
    "homepage": "https://wordpress.org/plugins/better-wp-security/",
    "names": [
        "wpackagist-plugin/better-wp-security"
    ],
    "versions": [
        "7.9.1"
    ],
    "source": {
        "type": "svn",
        "url": "https://plugins.svn.wordpress.org/better-wp-security/",
        "reference": "tags/7.9.1"
    },
    "dist": {
        "type": "zip",
        "url": "https://downloads.wordpress.org/plugin/better-wp-security.7.9.1.zip",
        "reference": null
    },
    "path": "/srv/www/cin/current/web/app/plugins/better-wp-security",
    "requires": {
        "composer/installers": "~1.0"
    }
}
```

**Comment by rarkins** [comment]
If it's in the lock file then that's easier for us - we can get it during the extract phase

**Comment by chrillep** [comment]
renovate works gr8 for creating PR's for type wordpress-plugin ❤️ .
What it currently does not do, is source changelogs.

* https://github.com/outlandishideas/wpackagist/issues/430#issuecomment-1589809932

Not sure if sourcing changelogs is a new issue tho

**Comment by rarkins** [comment]
Please start a new discussion

**Comment by chrillep** [comment]
> Please start a new discussion

https://github.com/renovatebot/renovate/discussions/24072
